text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,482 @@
+| `deliver`      | One or more messages are passed to a consumer. |
+| `settle`       | One or more messages are settled. |
+
+For further details about each of those operations refer to the [section about trace structure](#trace-structure).
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. `INTERNAL` otherwise. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+| `settle`       | `INTERNAL` |
+
+Setting span kinds according to this table ensures, that span links between
+consumers and producers always go from a `PRODUCER` span on the producer side to
+a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
+linked traces without the need for additional semantic hints.
+
+### Span relationships
+
+#### Producer spans
+
+"Publish" spans SHOULD be created for operations of providing messages for
+sending or publishing to an intermediary. A single "Publish" span can account
+for a single message, or for multiple messages (in the case of providing
+messages in batches). "Create" spans MAY be created. A single "Create" span
+SHOULD account only for a single message.
+
+If a "Create" span exists for a message, its context SHOULD be attached to
+the message. If no "Create" span exists, the context of the related "Publish"
+span SHOULD be attached to the message.
+
+#### Consumer spans
+
+##### Push-based scenarios
+
+"Deliver" spans SHOULD be created for operations of passing messages to the
+application when those operations are not initiated by the application
+code.
+
+##### Pull-based scenarios
+
+"Receive" spans SHOULD be created for operations of passing messages to the
+application when those operations are initiated by the application code.
+
+##### General considerations
+
+"Deliver" or "Receive" spans MUST NOT be created for messages which are not
+forwarded to the application, but are pre-fetched or cached by messaging
+libraries or SDKs.
+
+A single "Deliver" or "Receive" span can account for a single message, for
+multiple messages (in case messages are passed for processing as batches), or
+for no message at all (if it is signalled that no messages were received).  For
+each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
+the "Create" span for the message.
+
+#### Settlement spans
+
+"Settle" spans or events SHOULD be created for every manually or automatically
+triggered settlement operation. A single "Settle" span can account for a
+single message or for multiple messages (in case messages are passed for
+settling as batches). For each message it accounts for, the "Settle" span
```

Comment:
actually you could even got more messages one by one (single message) and then using a settlement for ALL of them. So it's not a batch but you are anyway using one settlement for more messages. Does the above sentence take this into account?

Reply:
I think you're referring to checkpoint-based settlement. 

For this case, I think we can't really expect to link to all messages that are settled by settling one message. The settlement span may have a sequence number attribute though, which indicates to what point the checkpoint was set.

Reply:
That's from the _Settlement_ section above:
> "Settle" spans may link to "Create" or "Publish" spans of the messages that are settled, however, for some settlement scenarios this is not feasible or possible.

In the end, this is a best-effort recommendation. Ideally, settlement spans have links to all the messages that are settled by the operation. However, in a considerable number of cases (and for checkpoint-based settlement in general) this is not feasible. That's why this is only put forth as a MAY, which is the weakest requirement level.

Reply:
Ok makes sense. Maybe it could be anyway useful to improve the [checkpoint-based settlement](https://github.com/open-telemetry/oteps/blob/main/text/trace/0173-messaging-semantic-conventions.md#checkpoint-based-settlement) section because something like that can happen with AMQP 1.0 (not just Kafka). The `disposition` frame allows you to specify `first` and `last` fields with delivery IDs you are settling. It means even more complex settlement than the checkpoint, because it could settle sparse range of messages.